roles/zuul/files/jobs/bonnyci.yaml

```diff
@@ -0,0 +1,43 @@
+- publisher:
+    name: bonnyci-logs
+    publishers:
+      - scp:
+          site: bonnyci-scp
+          files:
+            - target: 'logs/$LOG_PATH'
+              source: '/opt/bonnyci/logs/**'
+              keep-hierarchy: true
+              copy-after-failure: true
+
+- job-template:
+    name: 'bonnyci-run-{pipeline}'
+    pipeline: '{pipeline}'
+    test_log_dir: /opt/bonnyci/logs
+    builders:
+      - zuul-git-prep
+      - shell: |
+            #!/bin/bash
+            if [[ ! -f .bonnyci/run.sh ]]; then
+                echo "ERROR: No test script found at .bonnyci/run.sh, skipping"
+                exit 1
+            elif [[ ! -x .bonnyci/run.sh ]]; then
+                echo "ERROR: Test script found at .bonnyci/run.sh but it is not executable"
+                exit 1
+            fi
+
+            export BONNYCI_TEST_PIPELINE={pipeline}
+            export BONNYCI_TEST_LOG_DIR={test_log_dir}
+
+            (cd .bonnyci/ && ./run.sh)
```

Comment:
The way I envision people using this (and I could be wrong!), would be to have .bonnyci be a self-contained thing, probably with a few helper libraries or scripts in there that the main run.sh would source or call.  By executing run.sh within the subdir, the main script can just assume the things it needs are accessible via a path relative to $0.  I kinda had devstack plugins in the back of my head when putting this together.  Happy to hear what others think.